### PR TITLE
🔒 Add strong password indicators

### LINKED ISSUE #632: Add password strength indicators to signup screen
## Description
Add password strength indicators. Check out Xamarin app to see what they are and how they behave. The specs can be found in the old issues here https://github.com/toggl/mobileapp/projects/106
## Dependencies
#631 
## Definition of done
- Password strength criteria are added and behave in the same way as in Xamarin

### PR DESCRIPTION
**Pull request by pdrj**
### Summary
Adds strong password indicators

### Technical considerations
I had to make the Password classes not data classes, otherwise the password value on login screen is initially set to "Invalid("")", aka the string representation of invalid 😄 

### Relationships

Closes #632 

### Pull Request checklist

- [x] The code does not hinder the app's performance
- [x] The code does not introduce regressions
- [x] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests
- [ ] One of our [live or file custom templates](https://github.com/toggl/android/tree/master/.idea/templates) was used in this PR

### Pointers
Validate it behaves the same as in xamarin app. 
Validate you can still log in.

### Review template

- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones

- [ ] The code considers light and dark theme
- [ ] The code considers RTL and does not mention left/right
- [ ] The code considers i18n
  - [ ] There are no hardcoded strings in the UI
  - [ ] All strings are included in every language we support OR there's an issue requesting new translations
- [ ] There are no hard-coded numbers referring to dimensions or font sizes
- [ ] The code considers UI elements that are similar and extracts their similarities into styles to avoid duplication

## Merge permissions
Anyone

### CODE REVIEW

File: onboarding/src/main/java/com/toggl/onboarding/login/domain/LoginReducer.kt [hunk 1]
```diff
@@ -36,9 +36,9 @@ class LoginReducer @Inject constructor(
                     is Email.Invalid -> noEffect()
                     is Email.Valid -> when (val password = currentState.password) {
                         is Password.Invalid -> noEffect()
-                        is Password.Valid -> {
+                        is Password.Valid, is Password.Strong -> {
```
Comment (heytherewill) [inline onboarding/src/main/java/com/toggl/onboarding/login/domain/LoginReducer.kt:39]: A Strong password is already valid, I don't think this change is even needed 🤔

File: onboarding/src/main/java/com/toggl/onboarding/signup/ui/SignUpFragment.kt [hunk 2]
```diff
@@ -36,6 +44,24 @@ class SignUpFragment : Fragment(R.layout.fragment_signup) {
...
+
+        val defaultCriteriaColor = ContextCompat.getColor(requireContext(), R.color.authentication_text)
+        val errorCriteriaColor = ContextCompat.getColor(requireContext(), R.color.error)
+
+        val isPasswordStrongChannel = Channel<Boolean>()
+
+        passwordStrengthWatcher = password.doAfterTextChanged { isPasswordStrongChannel.offer(it.representsStrongPassword()) }
+
+        isPasswordStrongChannel
+            .consumeAsFlow()
```
Comment (heytherewill) [inline onboarding/src/main/java/com/toggl/onboarding/signup/ui/SignUpFragment.kt:56]: Why is this at all needed when the password already is in the state? 🤔 

```suggestion
        store.state
            .map { it.password is Password.Strong }
```

File: onboarding/src/main/java/com/toggl/onboarding/login/domain/LoginReducer.kt (hunk 1 above)
Comment (pdrj) [reply onboarding/src/main/java/com/toggl/onboarding/login/domain/LoginReducer.kt:39]: That was my understanding as well, but I'm pretty sure I couldn't login with a strong pass last week 😕 Today I can though, so I'll roll this change back 😄

File: models/src/main/java/com/toggl/models/validation/Password.kt [hunk 3]
```diff
@@ -3,13 +3,13 @@ package com.toggl.models.validation
 sealed class Password {
     abstract val password: String
 
-    data class Invalid internal constructor(override val password: String) : Password()
+    class Invalid internal constructor(override val password: String) : Password()
     open class Valid internal constructor(override val password: String) : Password() {
         override fun equals(other: Any?) = other is Valid && other.password == password
 
         override fun hashCode() = password.hashCode()
```
Comment (heytherewill) [inline models/src/main/java/com/toggl/models/validation/Password.kt:10]: I think the equality comparison needs to be on the strong password as well

Comment (pdrj) [reply models/src/main/java/com/toggl/models/validation/Password.kt:10]: https://toggl.slack.com/archives/CD40QCV25/p1596452350195700

**Pull request merged**